### ISSUE TITLE
SymPy formulas are not rendered correctly

### ISSUE THREAD

**Issue by holomorphism** [fonsp/Pluto.jl#488]
After the commit 1aa22b28bd48e1dd4dda3fdc1e7491c1f6cb94dd (#381), SymPy formulas are not rendered correctly.

before:
![ok](https://user-images.githubusercontent.com/3076585/94217599-60b79580-ff1d-11ea-9f23-1170b4704cf2.png)

after:
![ng](https://user-images.githubusercontent.com/3076585/94217605-64e3b300-ff1d-11ea-95b9-3bacf7306a51.png)

Thank you for the amazing software!

**Comment by fonsp** [comment]
Can you post the result of:
```julia
repr(MIME"text/latex"(), Sym("x"))
```

**Comment by holomorphism** [comment]
Sure.
![ng1](https://user-images.githubusercontent.com/3076585/94279474-1e2aa300-ff87-11ea-83b9-32e7b99304df.png)

**Comment by stillyslalom** [comment]
I think the underlying issue is MathJax's inability to mix display-mode equations and text. See [this](https://docs.mathjax.org/en/v2.7-latest/tex.html#differences), 

> First and foremost, the TeX input processor implements only the math-mode macros of TeX and LaTeX, not the text-mode macros. MathJax expects that you will use standard HTML tags to handle formatting the text of your page; it only handles the mathematics.

Given that, probably best to roll back #381 and provide good documentation on how to properly mix TeX and text.
![image](https://user-images.githubusercontent.com/4324760/95646279-6a4c0c00-0a8c-11eb-99aa-87d7bb456e0f.png)

**Comment by fonsp** [comment]
MathJax is able to mix text and math, as your screenshot demonstrates

**Comment by stillyslalom** [comment]
Yes, but the amsmath `equation*` environment (used by SymPy.jl for LaTeX repr) can't be wrapped with `\text`. It can be worked around by overriding https://github.com/JuliaPy/SymPy.jl/blob/b9028048dc2d8e8681ef3837c8232b8b262d0f3d/src/types.jl#L76 to change the output style from `equation*` to `inline`.
![image](https://user-images.githubusercontent.com/4324760/95659249-5e406880-0ae5-11eb-85b2-0adedec0ffde.png)

**Comment by fonsp** [comment]
> can't be wrapped
Are you referring to LaTeX spec or is this a limitation of MathJax?

**Comment by stillyslalom** [comment]
LaTeX spec. When I paste `\text{\begin{equation*} x \end{equation*}}` into TeX, I get "LaTeX Error: Bad math environment delimiter."
                                   
> Your command was ignored.
> Type  I <command> <return>  to replace it with another command,
> or  <return>  to continue without it.

...but it works fine without the `\text{...}` wrapper.

**Comment by stillyslalom** [comment]
For now, I'll work around this with
```
Base.show(io::IO, ::MIME"text/latex", x::SymPy.SymbolicObject) = print(io, sympy.latex(x, mode="inline"))
```
(posted here for easy copy-pasting)

**Comment by jtrakk** [comment]
That workaround doesn't quite work on some objects.

![image](https://user-images.githubusercontent.com/43392409/96390528-f4216680-1169-11eb-9f3a-2ae260506334.png)

**Comment by holomorphism** [comment]
@jtrakk How about the following code for `Matrix{Sym}` (for now)?

```julia
function Base.show(io::IO, ::MIME"text/latex", A::Union{AbstractVector{T},AbstractMatrix{T}}) where T <: SymPy.SymbolicObject
    rows = map(eachrow(A)) do row join(sympy.latex.(row),"&") end
    latex = "\$\\left[\\begin{array}{$("c"^size(A,2)))}" * join(rows,"\\\\") * "\\end{array}\\right]\$"
    print(io, latex)
end
```
This function also includes support for `Vector{Sym}`, but this is not reflected in Pluto (due to the way Pluto displays vector types).

I wrote this function by modifying the following code.  
<https://github.com/JuliaPy/SymPy.jl/blob/b9028048dc2d8e8681ef3837c8232b8b262d0f3d/src/types.jl#L78>

**Comment by fonsp** [comment]
Linking #623

**Comment by ambiso** [comment]
A very similar issue seems to pop up with Nemo's finite field representation:

![image](https://user-images.githubusercontent.com/3750466/99158027-9d5a6e80-26ce-11eb-8146-a07c993d288d.png)

The generated LaTeX seems to work, though:

![image](https://user-images.githubusercontent.com/3750466/99158038-b82ce300-26ce-11eb-87da-fb6d227460f9.png)

**Comment by fonsp** [comment]
Also see https://github.com/fonsp/Pluto.jl/issues/678#issue-743233421

**Comment by cenmir** [comment]
Workaround works, although I didn't like how some of the fractions were omitted so using the [SymPy docs](https://docs.sympy.org/latest/modules/printing.html#module-sympy.printing.latex) I found out about  `fold_short_frac`. It is set to true by default when the sympy.latex mode is set to inline. Testing setting it to false fixed the issue.

`Base.show(io::IO, ::MIME"text/latex", x::SymPy.SymbolicObject) = print(io, sympy.latex(x, mode="inline", fold_short_frac=false))`

Turns this output
![](https://i.imgur.com/WzshOuc.png)
into this
![](https://i.imgur.com/8DWVyVA.png)

Thanks for the help and an amazing software!

**Comment by holomorphism** [comment]
BTW, I wrote the following `mklatex` function to combine latex expressions in markdown.

```julia
begin
	function mklatex(xs...; inline=false)
		latex = strip(join(_latex.(xs), " "))
		if inline
			Markdown.LaTeX(latex)
		else
			Markdown.MD(Markdown.LaTeX(latex))
		end
	end

	if @isdefined SymPy
		_latex(x::SymPy.SymbolicObject) = sympy.latex(x)  # mode="plain"
	end

	_latex(x::AbstractString) = strip(x,'$')  # for LaTeXString

	function _latex(x::Rational)
		n, d = numerator(x), denominator(x)
		d == 1 ? string(n) : (n<0 ? "-" : "") * "\\frac{$(abs(n))}{$d}"
	end

	function _latex(x)
		if showable(MIME("text/latex"), x)
			repr("text/latex", x)
		else
			string(x)
		end
	end

	function _latex(x::Union{AbstractVector,AbstractMatrix})
		rows = map(eachrow(x)) do row join(_latex.(row),"&") end
		inner = join(rows, "\\\\")
		"\\left[\\begin{array}{$("c"^size(x,2))}" * inner * "\\end{array}\\right]"
	end
end
```
With it, you can do something like the following.

![latex](https://user-images.githubusercontent.com/3076585/100542970-bea38900-3290-11eb-8b23-eda4c8a9621f.png)

Isn't it useful? :wink:

**Comment by cenmir** [comment]
I really would like to be able to overload the Base.show function to generate
![](https://i.imgur.com/n1q1wwR.png)
from a column vector e.g., `x=[1,2]` (Array{T,1}).
I managed to get it working with an Array{T,2}. Feels like It's the last piece of the puzzle to really make this look nice for my students.

**Comment by holomorphism** [comment]
@fonsp 
I noticed that in a recent version of Pluto (v0.12.20), this problem has been fixed (thanks!).
Should the issue be closed?

However, the generated latex expression still contains redundancy, as shown below:
![01](https://user-images.githubusercontent.com/3076585/108491410-365cdf00-72e7-11eb-8742-37407ddf1006.png)

`\begin{equation*}...\end{equation*}` is from SymPy, and `\text{$...$\n}` is from Pluto.

I think it would be better to have a more concise latex output.
(Especially since I sometimes use MathJax's "copy to clipboard" feature).

Also, I haven't checked if this problem has been fixed for packages other than SymPy.

Other examples of latex representations:
![02](https://user-images.githubusercontent.com/3076585/108491439-41b00a80-72e7-11eb-9481-aa20c0c1b382.png)

**Comment by fonsp** [comment]
# How to contribute to this issue

To fix this, taking all packages into account, we need a **notebook with formulas** from multiple affected packages (including Markdown). Also take a look at the test cases written in #381

**Comment by eugene-shvarts** [comment]
i'm using Symbolics.jl, and can't get a matrix to pretty-print on evaluation. The culprit looks to be the `\text{}` tags surrounding the contents, because an L-string submitted without it works.
![Screen-Shot-2021-03-17-at-10 43 34-PM](https://user-images.githubusercontent.com/5599348/111704196-f876ac00-87fb-11eb-882e-aed27e7999f5.png)

Rendering directly with the L-string solves the problem, so if `D` is a matrix rendering improperly, then `L"%$(latexify(D))"` will render correctly.

**Comment by ChrisRackauckas** [comment]
The problem is that https://github.com/fonsp/Pluto.jl/pull/381 is wrong. It seems like LaTeXStrings.jl needs special handling, but that special handling is being applied to all LateX inputs instead of just that one package. If it's removed then SymPy and Symbolics.jl are just fine. I would think that the proper thing to do is to remove that handling and instead fix the LaTeXStrings.jl mime type dispatch

**Comment by stillyslalom** [comment]
The root of the problem is that MathJaX is meant to render math, not text, and we're wrapping every `MIME"text/latex"` object from every package with `\text`, which is not a generally-valid transformation for LaTeX. As Chris said, the original 'fix' is too broad in scope, and has caused trouble for many other LaTeX-friendly packages.

**Comment by zmoon** [comment]
MathJax is math-first but LaTeX (and presumably the `text/latex` MIME type) is text-first. My original implementation added `\text{}` only where necessary but that was rejected at the time for being too complex.

**Comment by stillyslalom** [comment]
Given that LaTeX in the context of a Pluto notebook will necessarily be rendered by MathJax, the opposite approach (rendering any non-mathmode content from a LaTeXString as plaintext) seems equally valid, but just as likely to wreak havoc in corner-cases.

**Comment by jtrakk** [comment]
It would be nice to follow some conventions about mathematical typesetting.

- https://nhigham.com/2016/01/28/typesetting-mathematics-according-to-the-iso-standard/
- http://journals.ieeeauthorcenter.ieee.org/wp-content/uploads/sites/7/IEEE-Math-Typesetting-Guide.pdf
- https://tug.org/TUGboat/tb18-1/tb54becc.pdf
- https://physics.nist.gov/cuu/pdf/typefaces.pdf

**Comment by VivaldoMendes** [comment]
After such a long issue, I have a kind of stupid question: how can I number an equation (written in LaTeX mode) in Pluto? 

The quality of the LaTeX output is fantastically displayed in Pluto (much, much better than in Jupyter), and we can write down LaTeX math code in different ways. But if we use Pluto for teaching, the numbering of equations becomes a vital issue. I can skip the crossref thing for now (at least, I still can do it manually), but the numbering of equations is essential. How can I do it?

Thanks.

**Comment by zmoon** [comment]
@VivaldoMendes I agree that things generally look prettier with the SVG renderer. To get equation numbers with MathJax 3 it has to [be enabled](http://docs.mathjax.org/en/latest/input/tex/eqnumbers.html) in the notebook's JavaScript. Pluto would have to provide an option to enable it afaik.

**Comment by VivaldoMendes** [comment]
I do not know much about the intricacies of JavaScript. I believe that the problem with `\label{}` is related to the cross-reference functionality of LaTeX, not with the auto-numbering. For a stater, we just need a way to introduce a number into a given equation written in LaTeX, and preferentially, an auto-numbering procedure.  I can use the trick `\quad  \quad  \quad  (1)` but it has three problems: it is inefficient, dangerous, and ugly.  I have tried some other alternatives, but none works. It looks too complicated to have many equations in a document and not having a way to call them by their label or number.

**Comment by zmoon** [comment]
@VivaldoMendes It seems like the best thing would be for Pluto to allow turning on the MathJax auto-numbering through configuration or a flag. Below is an example of how it could look. But #678 would have to be fixed too.

<details>
<summary>Example</summary>

Without auto-numbering enabled:
![image](https://user-images.githubusercontent.com/15079414/115937843-e5907080-a466-11eb-85c5-e15f4bb657f2.png)

With auto-numbering enabled:
![image](https://user-images.githubusercontent.com/15079414/115937821-d6a9be00-a466-11eb-8d39-9e9c27959368.png)

Code:
```html
<!DOCTYPE html>
<html>
<head>
<title>MathJax TeX Test Page</title>
<script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
<script type="text/javascript" id="MathJax-script" async
  src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-svg-full.js">
</script>


<!-- Enable equation numbering (and thus numbered references) -->
<script>
window.MathJax = {
  tex: {
    tags: 'ams'
  }
};
</script>

</head>
<body>
$$
\begin{equation}
y = m x + b

\label{eq:line}
\end{equation}
$$

$$
\begin{equation}
e^{i \pi} = -1
\end{equation}
$$

<p>
Equation \(\eqref{eq:line}\) is an equation for a line, which has reference \(\ref{eq:line}\).

The second equation does not have a label.
</body>
</html>

```
</details>

**Comment by VivaldoMendes** [comment]
@zmoon Thanks for your contribution. 

It looks exactly what is needed in Pluto.  Until it remains to be implemented in Pluto, is there a workaround to make this operational in the current Pluto version? If there is a way, it looks beyond what my skills can accomplish by now.

**Comment by feanor12** [comment]
Would something like this work as a workaround? So far it works for me.
```julia
Base.show(io::IO,::MIME"text/html",x::Num) = show(io,MIME("text/latex"),x)
```

**Comment by ChrisRackauckas** [comment]
Can we just fix Pluto here and do the workaround on LaTeXStrings? That would be the proper way to fix this.

**Comment by fonsp** [comment]
Fixed! Thanks @zmoon 

Test notebook is here: https://github.com/fonsp/disorganised-mess/blob/main/latex%20tests.jl

This broke support for mixed text-math LaTeXString (`L"hello $world$"`), but I was never very excited for those in the first place :)

**Comment by ChrisRackauckas** [comment]
Fantastic!

**Comment by fonsp** [comment]
Thanks for the help @ChrisRackauckas !

If anyone finds more issues with LaTeX rendering in Pluto, feel free to open a new issue!
